### PR TITLE
Add Prune API

### PR DESCRIPTION
**Pull request by sooahleex**
### Summary

- Ticket no.112037
- Support python api, project cli, cli for prune

### How to test

I added some cli and unit tests about this feature.

### Checklist

- [X] I have added unit tests to cover my changes.​
- [X] I have added integration tests to cover my changes.​
- [X] I have added the description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/datumaro/blob/develop/CHANGELOG.md).​
- [ ] I have updated the [documentation](https://github.com/openvinotoolkit/datumaro/tree/develop/docs) accordingly

### License

- [ ] I submit _my code changes_ under the same [MIT License](https://github.com/openvinotoolkit/datumaro/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

File: src/datumaro/cli/commands/prune.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        default=True,
+        help="""
+            Include this action as a project build step.
+            If true, this operation will be saved in the project
+            build tree, allowing to reproduce the resulting dataset later.
+            Applicable only to main project targets (i.e. data sources
+            and the 'project' target, but not intermediate stages)
+            (default: %(default)s)
+            """,
+    )
```
Comment (vinnamkim) [inline src/datumaro/cli/commands/prune.py]: Are those things being used for now?

File: src/datumaro/cli/commands/prune.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        prune_command: [
+            "target",
+            "method",
+            "ratio",
+            "hash_type",
+            "project_dir",
+            "model",
+            "dst_dir",
+        ]
+    }
```
Comment (vinnamkim) [inline src/datumaro/cli/commands/prune.py]: Is this being used for now?

File: src/datumaro/cli/commands/prune.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    parser.add_argument(
+        "-r", "--ratio", type=float, dest="ratio", help="How much to remain dataset after pruning"
+    )
+    parser.add_argument(
+        "--hash-type",
+        type=str,
+        dest="hash_type",
+        default="img",
+        help="Hashtype to extract feature from data information between image and text(label)",
+    )
```
Comment (vinnamkim) [inline src/datumaro/cli/commands/prune.py:59]: This argument seems not being passed in L130.

File: src/datumaro/components/prune.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+    def __init__(
+        self,
+        *datasets: Dataset,
+        cluster_method: str = "random",
+        hash_type: str = "img",
+    ) -> None:
+        """
+        Prune make a representative and manageable subset.
+        """
+        self._dataset = datasets[0] if isinstance(datasets, tuple) else datasets[0]
```
Comment (vinnamkim) [inline src/datumaro/components/prune.py]: Why do you take a sequence of datasets in L204 as the input argument if you want to store just one of them?

File: src/datumaro/components/prune.py [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+
+        self._prepare_data()
+
+    def _prepare_data(self):
+        if self._hash_type == "txt":
+            category_dict = self._prompting()
+
+        if self._cluster_method == "random":
+            self._item_list = list(self._dataset)
+        else:
```
Comment (vinnamkim) [inline src/datumaro/components/prune.py]:
```suggestion
        if self._cluster_method == "random":
            self._item_list = list(self._dataset)
            return
```

File: src/datumaro/components/prune.py [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+        for dataset_to_infer in datasets_to_infer:
+            if dataset_to_infer:
+                dataset_to_infer.run_model(self.model, append_annotation=True)
+        for dataset, dataset_to_infer in zip(datasets, datasets_to_infer):
+            updated_items = [
+                dataset.get(item.id, item.subset).wrap(annotations=item.annotations)
+                for item in dataset_to_infer
+            ]
+            dataset.update(updated_items)
+        return datasets
```
Comment (vinnamkim) [inline src/datumaro/components/prune.py]: Please make a common ancestor for `Explorer` and `Pruner` to remove duplicate functions.

File: src/datumaro/components/prune.py [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+        }
+
+        selected_items, dist_tuples = method[self._cluster_method](
+            ratio=ratio,
+            num_centers=self._num_centers,
+            labels=self._labels,
+            database_keys=self._database_keys,
+            item_list=self._item_list,
+            source=self._dataset,
+        )
```
Comment (vinnamkim) [inline src/datumaro/components/prune.py]: Please make a common ancestor having abstractmethod for those methods. It also should have correct typing for the input arguments.

File: src/datumaro/cli/commands/prune.py (hunk 3 above)
Comment (sooahleex) [reply src/datumaro/cli/commands/prune.py:59]: Remove this. For model, it will be updated later PR.

File: src/datumaro/components/prune.py (hunk 4 above)
Comment (sooahleex) [reply src/datumaro/components/prune.py]: Fix to get one dataset.

File: src/datumaro/cli/commands/prune.py (hunk 1 above)
Comment (sooahleex) [reply src/datumaro/cli/commands/prune.py]: Remove this. For model, it will be updated later PR.

File: src/datumaro/components/prune.py (hunk 6 above)
Comment (sooahleex) [reply src/datumaro/components/prune.py]: Make `HashInference` for `Explorer` and `Pruner`. If it is not proper, please let me know.

File: src/datumaro/components/prune.py (hunk 7 above)
Comment (sooahleex) [reply src/datumaro/components/prune.py]: I made `PruneBase` and override `base` for each method. I could not cover different input arguments for each method. If you have any suggestions on this part, I'd appreciate it if you could let me know.

File: src/datumaro/components/prune.py [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+        diff_num_item_list = np.argsort(cluster_weights - norm_cluster_num_item_list)
+        for diff_idx in diff_num_item_list[: abs(remaining_items)]:
+            norm_cluster_num_item_list[diff_idx] -= 1
+
+    return norm_cluster_num_item_list.tolist()
+
+
+class PruneBase(ABC):
+    @abstractmethod
+    def base(self):
```
Comment (vinnamkim) [inline src/datumaro/components/prune.py]:
```suggestion
    def base(self, ratio, num_centers, labels, database_keys, item_list, source) -> Tuple[List[DatasetItem], Optional[???]]:
```
Also add typing for the function input arguments

File: src/datumaro/components/explorer.py [hunk 9]
```diff
@@ -10,28 +10,40 @@
...
+        for dataset_to_infer in datasets_to_infer:
+            if dataset_to_infer:
+                dataset_to_infer.run_model(self.model, append_annotation=True)
+        for dataset, dataset_to_infer in zip(datasets, datasets_to_infer):
+            updated_items = [
+                dataset.get(item.id, item.subset).wrap(annotations=item.annotations)
+                for item in dataset_to_infer
+            ]
+            dataset.update(updated_items)
+        return datasets
```
Comment (vinnamkim) [inline src/datumaro/components/explorer.py]: Move this class implementation to a distinct source file such as `base.py`. Then, create a directory under `components/algorithms` and move it and the others (`components/explorer.py`, ...) to the created directory.

File: src/datumaro/util/hashkey_util.py
Comment (vinnamkim) [inline src/datumaro/util/hashkey_util.py:1]: This file also can be moved to the dedicated directory created above.

File: tests/integration/cli/test_prune.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+from datumaro.components.media import Image
+from datumaro.util.scope import scope_add, scoped
+
+from ...requirements import Requirements, mark_requirement
+
+from tests.utils.test_utils import TestDir
+from tests.utils.test_utils import run_datum as run
+
+
+class PruneTest(TestCase):
```
Comment (vinnamkim) [inline tests/integration/cli/test_prune.py]: Use `pytest` for the newly created tests

File: tests/unit/test_prune.py [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+from datumaro.components.media import Image
+from datumaro.components.prune import Prune, match_num_item_for_cluster
+from datumaro.plugins.data_formats.datumaro.exporter import DatumaroExporter
+
+from ..requirements import Requirements, mark_requirement
+
+from tests.utils.test_utils import TestDir
+
+
+class PruneTest(TestCase):
```
Comment (vinnamkim) [inline tests/unit/test_prune.py]: Use `pytest` for the newly created tests

Comment (sooahleex) [reply tests/unit/test_prune.py]: Fix in [4258c20](https://github.com/openvinotoolkit/datumaro/pull/1058/commits/4258c2060f46c723b1ecd46ce603b601bee77ade).

File: tests/integration/cli/test_prune.py (hunk 10 above)
Comment (sooahleex) [reply tests/integration/cli/test_prune.py]: Fix in [4258c20](https://github.com/openvinotoolkit/datumaro/pull/1058/commits/4258c2060f46c723b1ecd46ce603b601bee77ade).

File: src/datumaro/components/explorer.py (hunk 9 above)
Comment (sooahleex) [reply src/datumaro/components/explorer.py]: Move `components/explorer.py`, `components/prune.py`, `components/hashkey_util.py` into `components/algorithms/hash_key_inference`.

File: tests/unit/test_prune.py [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+
+        <b>Expected results:</b>
+        Pruned dataset that each subset contains one datasetitem.
+
+        <b>Steps</b>
+        1. Prepare dataset with each subset contains same images.
+        2. Set Prune and try get_pruned set method as ndr to extract representative subset.
+        3. Check whether each subset contains one datasetitem.
+        """
+        with TestDir() as test_dir:
```
Comment (vinnamkim) [inline tests/unit/test_prune.py]:
```suggestion
    def test_prune_ndr(self, fxt_dataset, test_dir):
        """
        <b>Description:</b>
        Check that pruned subset with ndr.

        <b>Input data:</b>
        Dataset with train and test subset that each datasetitem consists of same images.

        <b>Expected results:</b>
        Pruned dataset that each subset contains one datasetitem.

        <b>Steps</b>
        1. Prepare dataset with each subset contains same images.
        2. Set Prune and try get_pruned set method as ndr to extract representative subset.
        3. Check whether each subset contains one datasetitem.
        """
```

**Review by vinnamkim — APPROVED** [review]
LGTM.

**Review by wonjuleee — APPROVED** [review]
LGTM!

**Pull request merged**